planner/core/expression_rewriter.go

```diff
@@ -1481,8 +1481,19 @@ func (er *expressionRewriter) inToExpression(lLen int, not bool, tp *types.Field
 	if allSameType && l == 1 && lLen > 1 {
 		function = er.notToExpression(not, ast.In, tp, er.ctxStack[stkLen-lLen-1:]...)
 	} else {
+		coll, err := expression.CheckAndDeriveCollationFromExprs(er.sctx, "IN", types.ETInt, args...)
+		er.err = err
+		if er.err != nil {
+			return
+		}
 		eqFunctions := make([]expression.Expression, 0, lLen)
 		for i := stkLen - lLen; i < stkLen; i++ {
+			if er.ctxStack[i].GetType().EvalType() == types.ETString && l == 1 {
```

Comment:
Well, I got confused by the condition `l == 1`, could you explain it and add some comments here?